### DISCUSSION TITLE
Seb JSON string

### DISCUSSION THREAD

**Discussion by anticoder69** [SafeExamBrowser/seb-win-refactoring#778] [Q&A]
I've been writing an internal platform for our students. A bit of an E-assessement platform. In order to validate the integrity of the SEB config, I'm looking to integrate the Config Key SEB provides.
Given the steps in the developer documentation of Seb, I've come up with the following SEB-JSON string:

`{
    "additionalDictionaries": [],
    "additionalResources": [],
    "allowApplicationLog": false,
    "allowAudioCapture": false,
    "allowBrowsingBackForward": false,
    "allowCustomDownUploadLocation": false,
    "allowDeveloperConsole": false,
    "allowDictation": false,
    "allowDictionaryLookup": false,
    "allowDisplayMirroring": false,
    "allowDownUploads": false,
    "allowedDisplayBuiltin": false,
    "allowedDisplayBuiltinEnforce": true,
    "allowedDisplaysIgnoreFailure": false,
    "allowedDisplaysMaxNumber": 1,
    "allowFind": false,
    "allowFlashFullscreen": false,
    "allowPDFPlugIn": false,
    "allowPDFReaderToolbar": false,
    "allowPreferencesWindow": false,
    "allowPrint": false,
    "allowQuit": true,
    "allowScreenSharing": false,
    "allowSiri": false,
    "allowSpellCheck": false,
    "allowSpellCheckDictionary": [],
    "allowSwitchToApplications": false,
    "allowUserAppFolderInstall": false,
    "allowVideoCapture": false,
    "allowVirtualMachine": false,
    "allowWlan": true,
    "audioControlEnabled": true,
    "audioMute": false,
    "audioSetVolumeLevel": true,
    "audioVolumeLevel": 58,
    "backgroundOpenSEBConfig": false,
    "blockPopUpWindows": false,
    "browserMessagingPingTime": 120000,
    "browserMessagingSocket": "ws://localhost:8706",
    "browserScreenKeyboard": false,
    "browserShowFileSystemElementPath": false,
    "browserURLSalt": true,
    "browserUserAgentMac": 1,
    "browserUserAgentWinDesktopMode": 1,
    "browserUserAgentWinDesktopModeCustom": "VIRGO_SEB_PORTAL_CONFIG_",
    "browserUserAgentWinTouchMode": 2,
    "browserUserAgentWinTouchModeIPad": "Mozilla/5.0 (iPad; CPU OS 11_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.3 Mobile/15E216 Safari/605.1.15",
    "browserViewMode": 1,
    "browserWindowAllowAddressBar": false,
    "browserWindowAllowReload": true,
    "browserWindowShowURL": 0,
    "chooseFileToUploadPolicy": 0,
    "clipboardPolicy": 1,
    "createNewDesktop": true,
    "detectStoppedProcess": false,
    "downloadAndOpenSebConfig": true,
    "downloadDirectoryOSX": "~/Downloads",
    "downloadPDFFiles": false,
    "embeddedCertificates": [],
    "enableAltEsc": false,
    "enableAltF4": false,
    "enableAltMouseWheel": false,
    "enableAltTab": false,
    "enableAppSwitcherCheck": false,
    "enableBrowserWindowToolbar": false,
    "enableChromeNotifications": false,
    "enableCtrlEsc": false,
    "enableEsc": true,
    "enableF1": false,
    "enableF10": true,
    "enableF11": false,
    "enableF12": false,
    "enableF2": false,
    "enableF3": true,
    "enableF4": true,
    "enableF5": true,
    "enableF6": true,
    "enableF7": true,
    "enableF8": true,
    "enableF9": false,
    "enableFindPrinter": false,
    "enableJava": false,
    "enableJavaScript": true,
    "enableLogging": true,
    "enableMiddleMouse": false,
    "enablePlugIns": true,
    "enablePrintScreen": false,
    "enablePrivateClipboard": false,
    "enableRightMouse": true,
    "enableSebBrowser": true,
    "enableStartMenu": false,
    "enableTouchExit": false,
    "enableURLContentFilter": false,
    "enableURLFilter": false,
    "enableWindowsUpdate": false,
    "enableZoomPage": false,
    "enableZoomText": false,
    "examKeySalt": "CcfaCR7fJ4uiVs+VriSGE/jpyrh7XVhHd9UZGriRL74=",
    "examSessionClearCookiesOnEnd": true,
    "examSessionClearCookiesOnStart": false,
    "examSessionReconfigureAllow": true,
    "examSessionReconfigureConfigURL": "https://www.leerlingen.secured.virgosecundair.be/*",
    "exitKey1": 2,
    "exitKey2": 10,
    "exitKey3": 5,
    "forceAppFolderInstall": false,
    "hashedAdminPassword": "1660EDDCF4E3A5CDB89D2CC80079493CB1BB71157FEA9F66C1F9535C583AF7DC",
    "hashedQuitPassword": "FD0ACFCD711CB807EE7AAFD269E58B3E015714561ACC0ADDCC6786E1979ADDA3",
    "hideBrowserWindowToolbar": false,
    "hookKeys": true,
    "ignoreExitKeys": true,
    "insideSebEnableChangeAPassword": false,
    "insideSebEnableEaseOfAccess": true,
    "insideSebEnableLockThisComputer": false,
    "insideSebEnableLogOff": false,
    "insideSebEnableNetworkConnectionSelector": true,
    "insideSebEnableShutDown": false,
    "insideSebEnableStartTaskManager": false,
    "insideSebEnableSwitchUser": false,
    "insideSebEnableVmWareClientShade": false,
    "killExplorerShell": false,
    "lockOnMessageSocketClose": true,
    "logDirectoryOSX": "~/Documents",
    "mainBrowserWindowHeight": "100%",
    "mainBrowserWindowPositioning": 1,
    "mainBrowserWindowWidth": "100%",
    "minMacOSVersion": 4,
    "monitorProcesses": true,
    "newBrowserWindowAllowAddressBar": false,
    "newBrowserWindowAllowReload": false,
    "newBrowserWindowByLinkBlockForeign": true,
    "newBrowserWindowByLinkHeight": "100%",
    "newBrowserWindowByLinkPolicy": 0,
    "newBrowserWindowByLinkPositioning": 1,
    "newBrowserWindowByLinkWidth": "1000",
    "newBrowserWindowByScriptBlockForeign": false,
    "newBrowserWindowByScriptPolicy": 2,
    "newBrowserWindowNavigation": false,
    "newBrowserWindowShowReloadWarning": false,
    "newBrowserWindowShowURL": 0,
    "openDownloads": false,
    "oskBehavior": 2,
    "permittedProcesses": [],
    "pinEmbeddedCertificates": false,
    "prohibitedProcesses": [
        {
            "active": true,
            "currentUser": true,
            "executable": "obs32.exe",
            "originalName": "obs32.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "word.exe",
            "originalName": "word.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "AA_v3.exe",
            "originalName": "AA_v3.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "ptoneclk.exe",
            "originalName": "ptoneclk.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "webexmta.exe",
            "originalName": "webexmta.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "MS-Teams.exe",
            "originalName": "MS-Teams.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Teams.exe",
            "originalName": "Teams.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "join.me.sentinel.exe",
            "originalName": "join.me.sentinel.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "join.me.exe",
            "originalName": "join.me.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "g2mstart.exe",
            "originalName": "g2mstart.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "g2mlauncher.exe",
            "originalName": "g2mlauncher.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Telegram.exe",
            "originalName": "Telegram.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "obs64.exe",
            "originalName": "obs64.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Zoom.exe",
            "originalName": "Zoom.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Element.exe",
            "originalName": "Element.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "slack.exe",
            "originalName": "slack.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "CamtasiaUtl.exe",
            "originalName": "CamtasiaUtl.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "CamRecorder.exe",
            "originalName": "CamRecorder.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "CamPlay.exe",
            "originalName": "CamPlay.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Camtasia_Studio.exe",
            "originalName": "Camtasia_Studio.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "CamtasiaStudio.exe",
            "originalName": "CamtasiaStudio.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Camtasia.exe",
            "originalName": "Camtasia.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Discord.exe",
            "originalName": "Discord.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "pcmontask.exe",
            "originalName": "pcmontask.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "RPCSuite.exe",
            "originalName": "RPCSuite.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "RPCService.exe",
            "originalName": "RPCService.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "RemotePCDesktop.exe",
            "originalName": "RemotePCDesktop.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "beamyourscreen-host.exe",
            "originalName": "beamyourscreen-host.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "AeroAdmin.exe",
            "originalName": "AeroAdmin.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Mikogo-host.exe",
            "originalName": "Mikogo-host.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "chromoting.exe",
            "originalName": "chromoting.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "vncserverui.exe",
            "originalName": "vncserverui.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "vncviewer.exe",
            "originalName": "vncviewer.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "vncserver.exe",
            "originalName": "vncserver.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "PCMonitorSrv.exe",
            "originalName": "PCMonitorSrv.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "TeamViewer.exe",
            "originalName": "TeamViewer.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "GotoMeetingWinStore.exe",
            "originalName": "GotoMeetingWinStore.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "g2mcomm.exe",
            "originalName": "g2mcomm.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "SkypeHost.exe",
            "originalName": "SkypeHost.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "SkypeApp.exe",
            "originalName": "SkypeApp.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "Skype.exe",
            "originalName": "Skype.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "word.exe",
            "originalName": "word.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "excel.exe",
            "originalName": "excel.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "notepad.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "wordpad.exe",
            "originalName": "wordpad.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "strwinclt.exe",
            "originalName": "strwinclt.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "mstsc.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "SRServer.exe",
            "originalName": "SRServer.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "remoting_host.exe",
            "originalName": "remoting_host.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "CiscoWebExStart.exe",
            "originalName": "CiscoWebExStart.exe",
            "os": 1,
            "strongKill": true
        },
        {
            "active": true,
            "currentUser": true,
            "executable": "CiscoCollabHost.exe",
            "originalName": "CiscoCollabHost.exe",
            "os": 1,
            "strongKill": true
        }
    ],
    "proxies": {
        "AutoConfigurationEnabled": false,
        "AutoDiscoveryEnabled": false,
        "ExceptionsList": [],
        "ExcludeSimpleHostnames": false,
        "FTPEnable": false,
        "FTPPassive": true,
        "FTPPort": 21,
        "FTPRequiresPassword": false,
        "HTTPEnable": false,
        "HTTPPort": 80,
        "HTTPRequiresPassword": false,
        "HTTPSEnable": false,
        "HTTPSPort": 443,
        "HTTPSRequiresPassword": false,
        "RTSPEnable": false,
        "RTSPPort": 554,
        "RTSPRequiresPassword": false,
        "SOCKSEnable": false,
        "SOCKSPort": 1080,
        "SOCKSRequiresPassword": false
    },
    "proxySettingsPolicy": 0,
    "quitURL": "https://leerlingen.secured.virgosecundair.be/authorization/shutdown.php",
    "quitURLConfirm": true,
    "quitURLRestart": false,
    "removeBrowserProfile": true,
    "removeLocalStorage": false,
    "restartExamPasswordProtected": false,
    "restartExamUseStartURL": true,
    "sebAllowedVersions": [
        "Win.3.5.min"
    ],
    "sebConfigPurpose": 0,
    "sebMode": 0,
    "sebServiceIgnore": false,
    "sebServicePolicy": 0,
    "sendBrowserExamKey": true,
    "setVmwareConfiguration": false,
    "showApplicationLogButton": false,
    "showInputLanguage": false,
    "showMenuBar": false,
    "showReloadButton": true,
    "showReloadWarning": true,
    "showSideMenu": false,
    "showTaskBar": true,
    "showTime": true,
    "startURL": "https://leerlingen.secured.virgosecundair.be/portaal.php",
    "startURLAppendQueryParameter": true,
    "taskBarHeight": 40,
    "touchOptimized": false,
    "URLFilterEnable": false,
    "URLFilterEnableContentFilter": false,
    "urlFilterRegex": true,
    "URLFilterRules": [],
    "urlFilterTrustedContent": true,
    "useAsymmetricOnlyEncryption": false,
    "useTemporaryDownUploadDirectory": false,
    "zoomMode": 1
}`

But when hashing this with SHA256, and appending this to the absolute URL and re-hashing with sha256, the config key doesn't match.

My hunch is that my JSON file isn't correct. Could anyone point/help me in the right direction?

My code: (PHP):
`$json_hash = hash('sha256', $json);
$configKey = hash('sha256', $url.$json_hash);`

Thanks in advance!

**Comment by dbuechel** [comment]
Can't really help you with the PHP side of things, but please note that the example JSON on our website e.g. has no whitespaces: https://safeexambrowser.org/developer/seb-config-key.html.
